File: bench/cpp/alloc.cpp
#include <cstdio>

struct Node {
    int val;
    Node* next;
};

int main() {
    constexpr int N = 5000000;

    Node* head = nullptr;
    for (int i = 0; i < N; i++) {
        Node* n = new Node{i, head};
        head = n;
    }

    int total = 0;
    for (Node* c = head; c; c = c->next) {
        total += c->val;
    }

    Node* c = head;
    while (c) {
        Node* nx = c->next;
        delete c;
        c = nx;
    }

    std::printf("alloc total = %d\n", total);
    return 0;
}

File: bench/go/alloc.go
package main

import "fmt"

type Node struct {
	val  int32
	next *Node
}

func main() {
	const N int32 = 5_000_000

	var head *Node = nil
	for i := int32(0); i < N; i++ {
		head = &Node{val: i, next: head}
	}

	var total int32 = 0
	for c := head; c != nil; c = c.next {
		total += c.val
	}

	fmt.Printf("alloc total = %d\n", total)
}
